### PR TITLE
net, setup: Integrate the unplug flow through netpod (and nmstate)

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:

Integrate the network interface unplug flow through the `netpod` setup flow (which uses the `nmstate` API).

This change also includes a fix to the networks that are processed on setup (see last commit).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
~~This PR depends on #10518, only the last 6 commits are relevant.~~

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/network/setup/netpod/netpod.go [hunk 1]
```diff
@@ -400,7 +436,7 @@ func (n NetPod) masqueradeBindingSpec(podIfaceName string, vmiIfaceIndex int, if
 
 func (n NetPod) setupNAT(desiredSpec *nmstate.Spec, currentStatus *nmstate.Status) error {
 	bridgeIfaceSpec := n.lookupMasquradeBridge(desiredSpec.Interfaces)
-	if bridgeIfaceSpec == nil {
+	if bridgeIfaceSpec == nil || bridgeIfaceSpec.State == nmstate.IfaceStateAbsent {
```
Comment (AlonaKaplan) [inline pkg/network/setup/netpod/netpod.go:439]: Masquerade bridge cannot be absent.

File: pkg/network/setup/netpod/netpod_test.go [hunk 2]
```diff
@@ -954,6 +979,417 @@ var _ = Describe("netpod", func() {
...
+
+			// testNet1 and testNet2 are not expected to exist anymore.
+			_, err = cache.ReadDomainInterfaceCache(&baseCacheCreator, "0", testNet1)
+			Expect(err).To(HaveOccurred())
+
+			_, err = cache.ReadDomainInterfaceCache(&baseCacheCreator, "0", testNet2)
+			Expect(err).To(HaveOccurred())
+		})
+
+		It("unplug secondary bridge binding network that is still in pending state", func() {
```
Comment (AlonaKaplan) [inline pkg/network/setup/netpod/netpod_test.go:1273]: Please add a test unplugging a "Started" net.

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/a1233f2b0530dec73d58f8aa78f54125d6b9d1c9..310f50fc2b5bf8d09c67e3c08f1401706a84b981): Answered review comments.

File: pkg/network/setup/netpod/state_test.go [hunk 3]
```diff
@@ -104,10 +115,27 @@ var _ = Describe("state", func() {
...
+		state := netpod.NewState(newConfigStateCacheStub(), nil)
+		nets := []v1.Network{{Name: netName}}
+		Expect(state.SetFinished(nets)).To(Succeed())
+
+		Expect(state.Delete(nets)).To(Succeed())
+
+		pending, started, finished, err := state.PendingStartedFinished(nets)
+		Expect(err).NotTo(HaveOccurred())
+
+		// On deletion, all networks cache are initialized back to "pending".
```
Comment (ormergi) [inline pkg/network/setup/netpod/state_test.go:136]: Does the cache removed entirely when the interface is no longer exist in the spec? (removed by the controller following uplug)

File: pkg/network/setup/netpod/netpod.go [hunk 4]
```diff
@@ -210,22 +238,30 @@ func (n NetPod) composeDesiredSpec(currentStatus *nmstate.Status) (*nmstate.Spec
...
 
 			if nmstate.AnyInterface(ifacesSpec, hasIP4GlobalUnicast) {
 				spec.LinuxStack.IPv4.ArpIgnore = pointer.P(procsys.ARPReplyMode1)
 			}
+
+			if iface.State == v1.InterfaceStateAbsent {
+				var filteredIfacesSpec []nmstate.Interface
+				for _, ifaceSpec := range ifacesSpec {
+					// Interfaces with no type are not owned by kubevirt, therefore not removed.
+					if ifaceSpec.TypeName != "" {
```
Comment (ormergi) [inline pkg/network/setup/netpod/netpod.go:257]: Eventually only interfaces that specified in the VMI spec are processed, regardless of its type, isnt it?

**Comment by EdDev** [comment]
[change](https://github.com/kubevirt/kubevirt/compare/310f50fc2b5bf8d09c67e3c08f1401706a84b981..32ab077c9362f62b2fae26bf430c08aba8205e78): Responded to the review comments.

**Review by ormergi — APPROVED** [review]
Thanks for the changes, great work!

**Pull request merged**